resources/js/tableofcontent.js

```diff
@@ -0,0 +1,75 @@
+/**
+ * Construct recurseData.toc as a JSON-like that contains each header's text
+ * anchor, and their list of child headers.
+ * 
+ * References:
+ * https://stackoverflow.com/questions/187619/is-there-a-javascript-solution-to-generating-a-table-of-contents-for-a-page
+ * @param elem DOM element
+ * @param recurseData Object with fields: { toc : Object, count : int, parent : Object }
+ */
+function recurseChild(elem, recurseData)
+{
+    if (elem !== "undefined") 
+    {
+        var parent = recurseData.parent;
+        if(elem.nodeName.match(/^H\d+$/) && elem.textContent)
+        {
+            var headerText = elem.textContent.trim();
+            var level = elem.nodeName.substr(1);
+            var anchor = "kiwix-toc-" + recurseData.count;
+            recurseData.count += 1;
+
+            var anchorElem = document.createElement("a");
+            anchorElem.id = anchor;
+
+            /* Wrap header content with something we can reference. */
+            elem.insertAdjacentElement("afterbegin", anchorElem);
+            
+            /* Recurse back until the correct level for insertion. */
+            while (parent.level !== 0 && level < parent.level)
+                parent = parent.parent;
+            
+            var headerEntry = {text : headerText.replace(/"/g, '\\"'), anchor: "#" + anchor, table: []};
+            if (level === parent.level && parent.level !== 0)
+            {
+                /* Same level as parent meaning we should be sibling with parent. */
+                parent.parent.table.push(headerEntry);
+                parent = {parent: parent.parent, table: headerEntry.table, level: level};
+            }
+            else
+            {
+                /* We are the children of parent. */
+                parent.table.push(headerEntry);
+                parent = {parent: parent, table: headerEntry.table, level: level};
+            }
```

Comment:
This seems more convoluted than it could be. But let's address the more important issues first.

Reply:
I made an attempt to refactor the code.